app/models/reservation.rb

```diff
@@ -17,4 +17,17 @@ class Reservation < ApplicationRecord # :nodoc:
   validates :checkin_at, presence: true
   validates :checkout_at, presence: true
   validates :book_code, presence: true, uniqueness: true
+
+  # Callbacks
+  before_create :assign_available_room, if: -> { room.status == 'unavailable' }
+
+  private
+
+  def assign_available_room
+    room = Room.find_by(status: 'available', category: self.room.category)
+
+    raise StandardError, "No available room found for category #{self.room.category}" if room.nil?
```

Comment:
This looks to me more like a validation rule instead of an error being raised here. Do you always want a reservation to have a room assigned, right?

Reply:
I second that. If you raise here your app is going to crash and user won't be able to understand the problem.

You should probably do something along those lines:
```suggestion
    errors.add(:status, message: "room is not available anymore")
```

Reply:
@YannIsCoding the point here is that the reservation should not be created if there's not an available room, that's why I choose raise an error. The way user will understand that it's vital to assign a available room, will be handled soon on the controller. 
The logic to define availability will take place on Room side, does it make sense for you?

Reply:
In both cases you won't create a reservation. However, with the validation error you will return detailed information of why the data is invalid, which you can easily handle in the front-end. In the other case you are raising an error with a message, which is harder to handle and does not reflect the real problem in the backend. If in the future you want to add additional rules for the reservation, you will need to define and raise more errors, while with validations you just need to add more rules, but the logic to handle those validation errors stays the same. This is why this other approach is better. Errors are not really designed to handle these sort of issues.

Reply:
When you raise an error like this, it actually breaks the applications. Ie you get the standard rails red screen that says, oops something went wrong, not a user viewable error. So what you want is to add an error message so that it returns this error message to the user. In another word, when you raise, the error is for the developer to see, when you add error message (like a validation error), then the message can be shown to the user.

Reply:
Super, thanks for the clarification guys! ❤️